plugins/tedge_apama_plugin/src/main.rs

```diff
@@ -50,123 +49,217 @@ const TEDGE_APAMA_PROJECT_DIR: &str = "/etc/tedge/apama/project";
-                fs::remove_dir_all(tedge_apama_project_path)?;
-                println!("Removal of existing project successful");
-            }
+    println!(
+        "Installing newly extracted project to {}",
+        TEDGE_APAMA_PROJECT_DIR
+    );
+    fs::create_dir_all(tedge_apama_project_path)?;
+    fs::rename(tmp_apama_project_path, tedge_apama_project_path)?;
+    println!("Installation of new project successful");
 
-            success
-        }
-    };
+    println!("Restarting apama to load the new project");
+    let status = run_cmd("service", "apama restart")?;
+    println!("Restart of apama service successful");
 
     Ok(status)
 }
 
-fn run_cmd(cmd: &str, args: &str) -> Result<ExitStatus, InternalError> {
+fn remove_project(_project_name: &str) -> Result<(), InternalError> {
+    let tedge_apama_project_path: &Path = Path::new(TEDGE_APAMA_PROJECT_DIR);
+
+    if tedge_apama_project_path.exists() {
+        println!("Stopping apama service");
+        run_cmd("service", "apama stop")?;
+        println!("Stopping apama service successful");
+
+        println!("Removing existing project at {}", TEDGE_APAMA_PROJECT_DIR);
+        fs::remove_dir_all(tedge_apama_project_path)?;
+        println!("Removal of existing project successful");
+    }
+
+    Ok(())
+}
+
+fn get_installed_monitors() -> Result<Vec<String>, InternalError> {
+    // Run `engine_inspect -m -r` command to list all monitors in raw format
+    let output = Command::new(APAMA_ENV_EXE)
+        .arg(ENGINE_INSPECT_CMD)
+        .arg("-m")
+        .arg("-r")
+        .stdin(Stdio::null())
+        .output()
+        .map_err(|err| InternalError::exec_error(ENGINE_INSPECT_CMD, err))?;
+    let output = String::from_utf8(output.stdout)?;
+
+    // The output contains monitor names and their instance counts separated by a space as follows:
+    // ```
+    // TedgeDemoMonitor 1
+    // TedgeTestMonitor 1
+    // ```
+    let mon_files = output
+        .lines()
+        // The first line of the output could "WARNING: JAVA_HOME not set" which is filtered out
+        .filter(|line| !line.starts_with("WARNING:"))
+        // The counts are filtered out too
+        .filter_map(|line| line.split_whitespace().next())
+        .map(|line| line.into())
+        .collect();
+
+    Ok(mon_files)
+}
+
+fn install_or_update_monitor(mon_name: &str, mon_file_path: &str) -> Result<(), InternalError> {
+    let installed_monitors = get_installed_monitors()?;
+    // If an existing monitor needs to be updated, older version needs to be removed first before installing the new one
+    if installed_monitors.contains(&mon_name.to_string()) {
+        remove_monitor(mon_name)?;
+        install_monitor(mon_file_path)
+    } else {
+        install_monitor(mon_file_path)
+    }
+}
+
+fn install_monitor(mon_file_path: &str) -> Result<(), InternalError> {
+    run_cmd(
+        APAMA_ENV_EXE,
+        format!("{} {}", ENGINE_INJECT_CMD, mon_file_path).as_str(),
+    )
```

Comment:
run_cmd returns error? if so why not handle it?. I dint get how this will be forwarded if there is no `?`

Reply:
Note the missing `;` at the end of the function call indicating to return the result of this function call directly.